File: Leetcode_面试经典150题/219-150.java
// #219 存在重复元素 II
// 哈希表
// 滑动窗口(C)
package Leetcode_repository.Leetcode_面试经典150题;
import java.util.Map;
import java.util.HashMap;

class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        // 使用哈希表建立键值对：值 -> 下表
        Map<Integer, Integer> map = new HashMap<>(); // Diamond Syntax
        for (int i = 0; i < nums.length; i++) {
            if (map.containsKey(nums[i]) && i - map.get(nums[i]) <= k) {
                return true;
            }
            map.put(nums[i], i);
        }
        return false;
    }
}

File: Leetcode_面试经典150题/219-150.c
// #219 存在重复元素
//

// 使用哈希表需要自己实现，代码复杂，建议使用 Java



// 不使用哈希表
// 该算法最差时间复杂度 O(n^2) 超时了
#include <stdbool.h>

bool containsNearbyDuplicate(int* nums, int numsSize, int k){
    if(k <= 0){
        return false;
    }
    int *window = malloc(sizeof(int) * (k + 1));
    int windowsize = 0; // 当前窗口大小
    // 检查窗口中是否有重复元素
    for (int i = 0; i < numsSize; i++){
        for (int j = 0; j < windowsize; j++){
            if(window[j] == nums[i]){
                free(window);
                return true;
            }
        }
        // 添加到滑动窗口
        if(windowsize < k){
            window[windowsize++] = nums[i];
        }
        else{
            for (int i = 1; i < windowsize; i++){
                window[i - 1] = window[i];
            }
            window[windowsize - 1] = nums[i];
        }
    }
    free(window);
    return false;
}